### PR TITLE
[Script][BUG] Fix signature malleability for t inputs in Sapling txes

### PR DESCRIPTION
**Pull request by random-zebra**
We are currently signing the transaparent inputs with `SIGVERSION_BASE`, thus we are not considering the sapData in the hash.

The functional test included (`sapling_malleable_sigs.py`) provides a demo for a basic trolling exploit (intercept a `t->z` tx to effectively burn the amount of PIV that was supposed to be shielded).

**Fix**: 
Always use SIGVERSION_SAPLING to sign txes with version >= TxVersion::SAPLING.
This required a bit of refactoring, in order to pass the proper `sigversion`. 
In the future we could simplify this flow, saving the sigversion directly inside the SignatureCreator.

The json data for `sighash_test` unit test has been updated to reflect the new hashes.

The `sapling_malleable_sigs` functional test, now failing after the fix, is updated to verify that the proper exception is thrown.

### CODE REVIEW

**Comment by furszy** [comment]
Concept ACK, awesome catch! 👌 .

**Review by Fuzzbawls — APPROVED** [review]
Looking good

utACK e5b8dbafaa4b726315360b76256df56153940788

File: src/sapling/saplingscriptpubkeyman.cpp [hunk 1]
```diff
@@ -671,6 +671,9 @@ isminetype SaplingScriptPubKeyMan::IsMine(const CWalletTx& wtx, const SaplingOut
 
 CAmount SaplingScriptPubKeyMan::GetCredit(const CWalletTx& tx, const isminefilter& filter, const bool fUnspent) const
 {
+    if (!tx.IsShieldedTx() || tx.sapData->vShieldedOutput.empty()) {
```
Comment (furszy) [inline src/sapling/saplingscriptpubkeyman.cpp:674]: `IsShieldedTx` is asking for `vShieldedOutput.empty` inside. Would be good if we add a new method `hasShieldedOutputs`.
(Same for the `GetDebit`, an `hasShieldedSpends`)

**Review by furszy — APPROVED** [review]
Core review ACK e5b8dbafaa4b726315360b76256df56153940788 .
You won a big beer for this one

Two reminders for separate PRs:
1) Code organization discussion for the `SigVersion` being allocated inside the transaction primitive file when the signature lives on top of it.
2) New `sighash.json` for v1 transactions.

**Pull request merged**